### PR TITLE
Update pattern-variables.md

### PR DESCRIPTION
**Pull request by alrz**
Prototype: https://github.com/dotnet/roslyn/pull/59702
Rendered: [pattern-variables.md](https://github.com/alrz/csharplang/blob/patch-12/proposals/pattern-variables.md)

### CODE REVIEW

**Comment by alrz** [comment]
@jcouv Does this address your comment https://github.com/dotnet/csharplang/pull/4592#discussion_r721676260?

**Comment by alrz** [comment]
>`Could we relax the redeclaration requirement in a switch section?`

Note: I think that could be done if we enforce the redeclaration in terms of definite-assignment instead of relying on a declarator. This way the switch example works as presented. That would also result in no error for something like `var x or 1`, but `x` will be left unassigned. I'll reflect this in a separate PR later.

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
> I think that could be done if we enforce the redeclaration in terms of definite-assignment instead of relying on a declarator. 

I had a similar thought earlier, I think it's a good change that would be well received by LDM.

**Comment by alrz** [comment]
I went ahead and made the change in this PR. @333fred @jcouv please take a look, thanks.

**Comment by 333fred** [comment]
This all seems much more complicated than I was envisioning for rules, particularly the in-pattern ones. What I was envisioning was simply: if a variable is not currently definitely assigned, it may be redeclared in a pattern. This would permit `a is T t || b is T t` because `t` is not definitely assigned in the false case, but would not permit `_ = a is T t; _ = b is T t;`

**Comment by alrz** [comment]
>`a is T t || b is T t`

If I recall correctly, LDM didn't support redeclaration across expression boundaries.  I can readd the open question on that.
>This all seems much more complicated than I was envisioning for rules

I only reverted the initial PR and made some changes (this currently mirrors the prototype). 
I'm not sure how this could be specced in simpler terms while maintaining the same behavior. Suggestions?

**Comment by 333fred** [comment]
> If I recall correctly, LDM didn't support redeclaration across expression boundaries. I can readd the open question on that.

Right, we were leary of the complex rules involved and concerned about double assignments. I think that my proposed simplification, where it's just "you can assign if it's not definitely assigned", is much simpler and alleviates the double assignment concerns.

> I'm not sure how this could be specced in simpler terms while maintaining the same behavior. Suggestions?

My main thought is to simply remove all the things about where variables are allowed to be redeclared in patterns, and simply state that variables are allowed to be redeclared by patterns if they are not definitely assigned and the types all line up. We might have to define what definite assignment means _inside_ a pattern, but I think that's much more straightforward than the existing wording of the rules around where in patterns variables are allowed to be redeclared.

**Comment by alrz** [comment]
Just to make sure we're on the same page, "you can assign if it's not definitely assigned" means pattern variables can be effectively redeclared *anywhere* within their scope; permitting `if (a is C c) ; else if (b is C c) ;` for example.

For binary patterns, we need to somehow flow the def assignment state from LHS so that the variables are considered assigned in the RHS (similar to expressions) to catch both `not (0, var x) or not (var x, 0)` or `(0, var x) and (var x, 0)` as errors. 

That doesn't prevent `_ = a is T t; _ = b is T t;` though, `t` could be unassigned when we enter the second statement.
Same for `case not var t: case var _ when e is var t:` because we've relaxed the redeclaration rule in the entire scope.

**Comment by 333fred** [comment]
> Just to make sure we're on the same page, "you can assign if it's not definitely assigned" means pattern variables can be effectively redeclared _anywhere_ within their scope; permitting `if (a is C c) ; else if (b is C c) ;` for example.

Yes.

> For binary patterns, we need to somehow flow the def assignment state from LHS so that the variables are considered assigned in the RHS (similar to expressions) to catch both `not (0, var x) or not (var x, 0)` or `(0, var x) and (var x, 0)` as errors.

Yes

> That doesn't prevent `_ = a is T t; _ = b is T t;` though, `t` could be unassigned when we enter the second statement. Same for `case not var t: case var _ when e is var t:` because we've relaxed the redeclaration rule in the entire scope.

Indeed. I don't especially care about that scenario: you can't use `t` anyway.

File: proposals/pattern-variables.md [hunk 1]
```diff
@@ -1,73 +1,96 @@
...
-    case (int x, 0) a when Use(x, a): // ok
-    case (0, int x) b when Use(x, b): // ok
-        Use(x); // ok
-        Use(a); // error; not definitely assigned
-        Use(b); // error; not definitely assigned
-        break;
-    ```
+### Variable redeclaration
+
+- Pattern variables are allowed to be redeclared within their scope if not already definitely assigned. These names can possibly reference either of variables based on the result of the pattern-matching at runtime.
```
Comment (alrz) [inline proposals/pattern-variables.md:58]: I think this should not include nested scoped, e.g.
```cs
if (e is C c) return;
if (e is C c) return; // allowed
{ if (e is C c) return; } // not allowed
```
Not completely sure if "within their scope" implies that.

Comment (333fred) [reply proposals/pattern-variables.md:58]: Perhaps `in the same scope`?

Comment (333fred) [reply proposals/pattern-variables.md:58]: It also would probably be worth moving it to an open questions section.

**Review by 333fred — COMMENTED** [review]
I like this version a lot more :). Just a couple small comments.

File: proposals/pattern-variables.md (hunk 1 above)
Comment (alrz) [reply proposals/pattern-variables.md]: There's a special error for that (CS0136) which I believe shouldn't be relaxed - in interaction with loops it gets too confusing. I think we should draw the line *somewhere* as the current rules are already too loose. Would like to see how all this play out.

Comment (alrz) [reply proposals/pattern-variables.md]: @333fred Done. Any chance this could be scheduled for ldm review? Thanks.

**Review by 333fred — APPROVED** [review]
@jcouv please take a look

**Comment by jcouv** [comment]
Good to merge.
@alrz We were planning to review this morning, but ended up not having time. Likely will get discussed next week.
